### PR TITLE
store: retry auth-related requests

### PR DESCRIPTION
**Pull request by stolowski**
Retry the requests from auth.go part of the store. Based on #2453 .

### CODE REVIEW

**Comment by pedronis** [comment]
is it worth having a helper like  retryRequestDecodeJSON for these?

**Comment by stolowski** [comment]
@pedronis Yes, I agree a helper is a good idea. I'm on it but it requires quite some refactoring. Perhaps it makes sense to land this as is to help with the network failures in the tests and I'll propose the refactored version soon.

File: store/auth.go [hunk 1]
```diff
@@ -80,45 +82,107 @@ func loginCaveatID(m *macaroon.Macaroon) (string, error) {
...
+		}
+		if result != nil {
+			return json.NewDecoder(resp.Body).Decode(result)
+		}
+		return nil
+	})
+}
+
+// retryRequest calls doRequest and decodes the response in a retry loop.
+func retryRequest(endpoint string, headers map[string]string, reader io.Reader, decode func(ok bool, resp *http.Response) error) (resp *http.Response, err error) {
```
Comment (pedronis) [inline store/auth.go]: don't we an issue with io.Reader about trying to read it again?  at least in store.go we ended up using []byte instead?

File: store/auth.go [hunk 2]
```diff
@@ -175,14 +234,9 @@ func requestDischargeMacaroon(endpoint string, data map[string]string) (string,
 		return "", fmt.Errorf(errorPrefix+"server returned status %d", resp.StatusCode)
 	}
 
-	dec := json.NewDecoder(resp.Body)
-	var responseData struct {
-		Macaroon string `json:"discharge_macaroon"`
-	}
-	if err := dec.Decode(&responseData); err != nil {
+	if err != nil {
```
Comment (pedronis) [inline store/auth.go]: this is already covered now

Comment (pedronis) [inline store/auth.go]: looking at the full file there's also another place that I don't see in the diff with a repeated check like this

File: store/auth.go (hunk 1 above)
Comment (pedronis) [inline store/auth.go]: some of the retry test should probably check that the body is correct in the retries too,

anyway given the usages here data []byte  would be fine afaict quickly

File: store/retry.go [hunk 3]
```diff
@@ -20,12 +20,45 @@
...
+	retry.Exponential{
+		Initial: 100 * time.Millisecond,
+		Factor:  2.5,
+	},
+))
+
+func maybeLogRetryAttempt(url string, attempt *retry.Attempt, startTime time.Time) {
+	if osutil.GetenvBool("SNAPPY_TESTING") || attempt.Count() > 1 {
+		delta := time.Since(startTime) / time.Millisecond
+		logger.Debugf("Retyring %s, attempt %d, delta time=%v ms", url, attempt.Count(), delta)
```
Comment (chipaca) [inline store/retry.go]: remember you can just pass a `time.Duration` to `%s` and it'll print e.g. `123ms` if that is the most appropriate unit for the duration.

File: store/retry.go [hunk 4]
```diff
@@ -20,12 +20,45 @@
...
+// Timeout value
+var defaultRetryStrategy = retry.LimitCount(5, retry.LimitTime(33*time.Second,
+	retry.Exponential{
+		Initial: 100 * time.Millisecond,
+		Factor:  2.5,
+	},
+))
+
+func maybeLogRetryAttempt(url string, attempt *retry.Attempt, startTime time.Time) {
+	if osutil.GetenvBool("SNAPPY_TESTING") || attempt.Count() > 1 {
```
Comment (pedronis) [inline store/retry.go]: it's an OR, it means we also log when there was no attempt in our tests

Comment (chipaca) [inline store/retry.go]: D'ouch

File: store/auth.go [hunk 5]
```diff
@@ -80,45 +82,112 @@ func loginCaveatID(m *macaroon.Macaroon) (string, error) {
...
+	})
+}
+
+// retryPostRequest calls doRequest and decodes the response in a retry loop.
+func retryPostRequest(endpoint string, headers map[string]string, data []byte, decode func(ok bool, resp *http.Response) error) (resp *http.Response, err error) {
+	var attempt *retry.Attempt
+	startTime := time.Now()
+	for attempt = retry.Start(defaultRetryStrategy, nil); attempt.Next(); {
+		var body io.Reader
+		if data != nil {
```
Comment (pedronis) [inline store/auth.go]: no need to do this check,

body := bytes.NewBuffer(data) would be ok (also for nil)

File: store/retry.go [hunk 6]
```diff
@@ -20,12 +20,45 @@
...
-	"gopkg.in/retry.v1"
+	"fmt"
 	"io"
 	"net"
 	"net/http"
+	"time"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"gopkg.in/retry.v1"
```
Comment (pedronis) [inline store/retry.go]: forgot this?

File: store/retry.go [hunk 7]
```diff
@@ -20,12 +20,44 @@
...
+
+func logRetryTime(startTime time.Time, url string, attempt *retry.Attempt, resp *http.Response, err error) {
+	var status string
+	delta := time.Since(startTime) / time.Millisecond
+	if err != nil {
+		status = err.Error()
+	} else if resp != nil {
+		status = fmt.Sprintf("%d", resp.StatusCode)
+	}
+	logger.Debugf("The retry loop for %s finished after %d retries, delta time=%v ms, status: %s", url, attempt.Count(), delta, status)
```
Comment (chipaca) [inline store/retry.go]: the comment about time.Duration applies here also :-D

**Pull request merged**